### ISSUE TITLE
MirageOS 4.0 draft release roadmap

### ISSUE THREAD

**Issue by hannesm** [mirage/mirage#994]
Based on discussions in Marrakesh (Oct 2018) etc., please let me try to draft a roadmap (please discuss by commenting in this issue - also in case items are missing, or you want to be in charge of a specific item, or an item should be out of scope for MirageOS 4.0). All errors and typos are my fault!

Being a library operating system we should avoid unnecessary blocking of (unrelated) projects. At least to me, its mainly a question when to tag and announce 4.0.

The main achievement from a unikernel developer perspective is to use dune instead of ocamlbuild, get rid of `pkg-config` (replaced by dune magic), and being able to automatically construct a monorepo for a unikernel (using duniverse) - which makes it easier to modify and adapt any of the code in a unikernel (not having to mess around with opam pins).

For users, a major stepping stone is persistent storage (via a block device and file system, git remote, … - using the same interface (mirage-kv)), monitoring, multiple network (and block) interfaces on all platforms, and improved documentation.

In addition, having a more clear maintainence story for our repositories would be great (see this PR
mirage/mirage-www#562), use [CODEOWNERS files](https://help.github.com/en/articles/about-code-owners).

Development tools
- [x] functoria uses dune mirage/functoria#171 mirage/functoria#167 (~~functoria 3.0 released~~) mirage/functoria#175 mirage/functoria#176 mirage/functoria#177
- [x] ~mirage dune #979~
- [ ] use variants for mirage-block-*
- [ ] use variants for mirage-net-*
- [ ] document dune and mirage workflow (see https://gist.github.com/TheLortex/f3d92db831b553f6e4eaf2982e3e6427)
- [x] Port existing mirage-skeleton to dunified mirage (see https://github.com/TheLortex/mirage-skeleton)

Mirage utility
- [ ] block device naming #890
- [x] network device naming; with multi-device solo5, have to pass `~group` around (see e.g.
https://github.com/hannesm/mirage-skeleton/commit/0b39e20e0a854986c12648c5fe8a7aead9b2857d#diff-3c75748e7e67dc7251b85561580907d8)
- [ ] adding maintainers; see mirage/mirage-www#562, https://help.github.com/en/articles/about-code-owners

Deployment
- [ ] Public cloud ??
- [x] Own infrastructure; https://github.com/hannesm/albatross
- [ ] Kubernetes (see https://github.com/keyfender/deployment/tree/master/k8s)

Library development
- [ ] Persistent storage; unclear state of wodan, solo5/solo5#325 solo5/solo5#329 solo5/solo5#330
- [x] Deprecate mirage-fs (use mirage-kv 2.0 instead)
- [x] ~~Deprecate io-page (or merge into cstruct?); maybe skip~~
- [x] ~~Deprecate mirage-console; maybe skip~~
- [x] Release metrics
- [x] Provide more metrics sources; mirage/mirage-solo5#43
- [x] Provide metrics reporters; mirage/metrics#28
- [x] Cstruct and capabilities (see mirage/ocaml-cstruct#237) - would be nice to integrate in all the libraries
- [x] irmin2 (and git2) releases mirage/irmin#658

Targets
- [x] e2e testing for solo5; https://github.com/mato/e2e-mirage-solo5
- [x] Solo5 seccomp (spt) solo5/solo5#310
- [x] Solo5 multi-device solo5/solo5#375
- [x] Mirage (and other) changes #992 for multi-device and seccomp
- [x] Defense in depth (solo5); W^X not easily possible on kvm (solo5/solo5#303); heap canaries mirage/ocaml-freestanding#48; no `-m-no-red-zone` solo5/solo5#292; ASLR solo5/solo5#304
- [x] OpenBSD pledge
- [x] FreeBSD capsicum solo5/solo5#366
- [x] Xen new modernised PVH-only stack, see #1159
- [ ] ESP32; rebased on dune + mirage, still uses espressif SDK

Documentation
- [ ] CI story; in progress
- [ ] fuzzing story; where to publish the documents?
- [ ] cut & paste examples (mdx); unclear whether mdx is already suitable for writing unikernels (two-stage compilation); would like to see example

Example applications
- [x] new DNS client, caching recursive and caching stub resolvers, authoritative server; needs further fixes and release https://github.com/mirage/ocaml-dns ; mirage/ocaml-conduit#290 (also https://github.com/roburio/unikernels esp. "future" branch) -- also https://hannes.nqsb.io/Posts/Summer2019
- [ ] EMail server; client released, awaiting user feedback
- [x] TLS reverse proxy (https://github.com/hannesm/tlstunnel, but as a unikernel!) -- there's https://github.com/roburio/unikernels/tree/tls/tlstunnel
- [x] CalDAV server; see https://github.com/roburio/caldav/tree/future - requires dns release, irmin2 -- also https://hannes.nqsb.io/Posts/Summer2019

**Comment by samoht** [comment]
I'm in favour of focusing the 4.0 release on the tooling only (e.g. mirage4 = mirage+dune). That's already a big change and that'll make the rest easier to release/maintain.

**Comment by mor1** [comment]
Generally agree with @samoht I think. Might it make sense to do a 3.x release containing just the Tooling tasks, and then push on with Library Development and Deployment tasks for 4.0? Will give a chance for issues with tooling changes to be surfaced while the rest of the universe remains somewhat stable.

(On the basis that presumably tooling changes are mostly transparent and don't require much in the way of code changes? Or am I assuming too much there?)

**Comment by dinosaure** [comment]
I agree too when it becomes a pain to live in both worlds (mirage without dune and mirage with dune). duniverse and the @TheLortex's work should be the next step. Currently, I tried to use this work on my side and it mostly works (some details still need a polishement).

It should be interesting (as a proof of concept) to move some others unikernels (like some from robur.io) to mirage+dune.

**Comment by hannesm** [comment]
I don't see much contradiction in there, though I'd suggest to have the `Development tools` and `Mirage utility` tasks done before a 4.0 release as well -- since the mirage tool semantics should be mostly stable during a major release cycle.

For the `mirage+dune` part, I still lack some idea about the concrete design (it seems to require this not yet released "duniverse" utility which is very similar to opam (caching, fetching sources, ...), and adds an overlay of packages (IIUC https://github.com/avsm/mirage-duniverse)), especially in respect to update workflows and discovering where the sources originate from are imho crucial.

I use various unikernels in production, and having their `app_info` (i.e. (conservative) list (including build dependencies as well, plus optional dependencies) of (transitive) dependencies with version numbers _as released in opam_) is pretty good for figuring out which unikernels I need to upgrade once I encounter (and fix) an issue in a library.

**Comment by dinosaure** [comment]
I understand your concerns where a replication of the unikernel should be important. The current status of MirageOS + dune and duniverse is not heavily linked (if we don't consider some details).

The `dune` file generated by the `mirage` tool expects in any way vendors. Then, the way to get them, currently, is to use `duniverse` which will only fetch repositories according the OPAM file (which is generated by `mirage` too).

I don't know so much about details on `duniverse` but when I spoke with @Julow, it's possible to use something else than `duniverse` currently (and it seems a goal to not heavily depends on `duniverse`). However, in reality, it seems that in some details, we need `duniverse`. From what I know, `duniverse` still continues to fetch only version (a `duniverse update` want to fetch the dev/trunk/master version).

My only concern currently is the mix between the OPAM environment which can provide some libraries and the `duniverse` environment which should provide the same library when it's not so clear which distribution `dune` will use to compile the unikernel.

**Comment by hannesm** [comment]
now with #1004 having some traction, there are some more API/interface changes that we may do for the next major version:
- should PCLOCK use Ptime.t instead of the int * int64? I'd be in favour to avoid the calls to `Ptime.v (PCLOCK.now_d_ps ())` -- this can be done in a backwards-compatible way by introducing `PCLOCK.now : unit -> Ptime.t`
- with Cstruct_cap (and the reduced opam packages), I'd like to PoC our network stack (mirage-net / mirage-protocols) to use these capabilities -- this is certainly a breaking change, but I'm all in favour thereof
- mirage-kv: I'd appreciate it to be functorised over `type key` -- since it turns out I like lots of flat KV-stores, and find the Mirage_kv.Key.t in these cases a bit over the top -- e.g. a caching web-proxy
- mirage-device: here I'm very uncertain about the errors -- are `` `Disconnected `` sounds fine (if we want to have lifecycles of devices -- do we? related is https://github.com/Solo5/solo5/issues/406), but `` `Unimplemented `` is not anything I can think of handling programmatically properly -- I agree with @cfcs that we need to take the xen devices into account as well

I'm happy to write up documentation about versioning of MirageOS interfaces once #1004 lands (basically every change (apart maybe from adding a binding -- but that can break code due to local opens) is a major version bump).

**Comment by hannesm** [comment]
From the janitor point of view, mirage-device (http://opam.ocaml.org/packages/mirage-device/) can be removed for MirageOS 4.0.0 (needs to be deprecated in a 3.x series, the interface is only used internally and not very useful). This requires the reverse dependencies being bumped as well.

**Comment by dinosaure** [comment]
> Public cloud ??

As far as I can tell and from my experiment, it's "easy" to deploy MirageOS on Google Cloud if you mean such service for public cloud. The story seems a bit more difficult for AWS but the main blocker seems about the tooling around such service instead of MirageOS itself.

Some articles explain how to deploy on `gcloud`:
- https://www.riseos.com/blog/2016/09/13/continuously-deploying-mirage-unikernels-to-google-compute-engine-using-circle-ci.html
- https://blog.container-solutions.com/all-about-unikernels-part-2-mirageos-and-rumprun

They are a bit outdated but still are true :+1:.

**Comment by adamsteen** [comment]
@hannesm When you say OpenBSD Pledge, what do you mean? The Solo5 hvt tender on OpenBSD is pledged, or are thinking something else?

**Comment by hannesm** [comment]
@adamsteen yes indeed that was the goal. I ticked it off, thanks for your comment (and your work on that)! :)

**Comment by hannesm** [comment]
I'm closing this issue - it served its purpose at the time being, but these days #1217 is more accurate for the actual release. If you find something that is worth keeping in the issue tracker, please open a new issue.
